torch/csrc/jit/passes/onnx/shape_type_inference.cpp

```diff
@@ -1143,18 +1143,89 @@ void SpecialPostProcess(Node* n) {
       // onnx Sequence type requires element type to be set.
       // If the list to insert is empty, we set the elem type by
       // looking at the tensor being inserted.
-      auto list_node = n->input(0)->node();
+      auto seq_node = n->input(0)->node();
       auto t_node = n->input(1)->node();
-      if (!list_node || list_node->kind() != prim::ListConstruct ||
-          list_node->inputs().size() != 0) {
-        break;
-      }
+      auto t_type = n->input(1)->type()->cast<TensorType>();
+
+      auto update_sequence_empty_dtype = [](Node* n, TensorTypePtr t_type) {
+        TORCH_INTERNAL_ASSERT(n && n->kind() == ::c10::onnx::SequenceEmpty);
+        TORCH_INTERNAL_ASSERT(t_type->scalarType().has_value());
+        auto scalar_type = t_type->scalarType().value();
+        auto onnx_type = ATenTypeToOnnxType(scalar_type);
+        n->i_(attr::dtype, onnx_type);
+        n->output()->setType(ListType::create(t_type));
+      };
+
+      auto find_sequence_empty = [](Value* input,
```

Comment:
Do we need lambda function here? Can we just write it as regular function outside `void SpecialPostProcess(Node* n)`?
Only see this `find_sequence_empty` used once. Same question for `find_sequence_empty_impl`

Reply:
The reason for lambda is exactly that because it's only used once, for `SequenceInsert`. If there are use cases then we can make it public and more general. The `find_sequence_empty_impl` is a trick to enable recursive lambda function. `find_sequence_empty` need to recursively call itself in case of nested blocks.